server/src/main/java/org/elasticsearch/action/support/StatsRequestLimiter.java

```diff
@@ -0,0 +1,121 @@
+import org.elasticsearch.common.util.concurrent.AdjustableSemaphore;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.common.util.concurrent.RunOnce;
+import org.elasticsearch.tasks.Task;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class guards the amount of stats requests a node can concurrently coordinate.
+ */
+public class StatsRequestLimiter {
+
+    public static final Setting<Integer> MAX_CONCURRENT_STATS_REQUESTS_PER_NODE = Setting.intSetting(
+        "node.stats.max_concurrent_requests",
+        100,
+        1,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final AdjustableSemaphore maxConcurrentStatsRequestsPerNodeSemaphore;
+    private final Map<String, StatsHolder> stats = new ConcurrentHashMap<>();
+
+    public StatsRequestLimiter(Settings settings, ClusterSettings clusterSettings) {
+        this.maxConcurrentStatsRequestsPerNodeSemaphore = new AdjustableSemaphore(
+            MAX_CONCURRENT_STATS_REQUESTS_PER_NODE.get(settings),
+            false
+        );
+        clusterSettings.addSettingsUpdateConsumer(MAX_CONCURRENT_STATS_REQUESTS_PER_NODE, this::setMaxConcurrentStatsRequestsPerNode);
+    }
+
+    private void setMaxConcurrentStatsRequestsPerNode(int maxConcurrentBoundedDiagnosticRequestsPerNode) {
+        this.maxConcurrentStatsRequestsPerNodeSemaphore.setMaxPermits(maxConcurrentBoundedDiagnosticRequestsPerNode);
+    }
+
+    /**
+     * Checks if executing the action will remain within the limits of the max concurrent requests the node can handle. If the limit is
+     * respected the action will be executed otherwise it will throw an EsRejectedExecutionException. The method keeps track of current,
+     * completed and rejected requests per action type.
+     */
+    public <Request, Response> void tryToExecute(
+        Task task,
+        Request request,
+        ActionListener<Response> listener,
+        TriConsumer<Task, Request, ActionListener<Response>> executeAction
+    ) {
+        StatsHolder statsHolder = stats.computeIfAbsent(task.getAction(), ignored -> new StatsHolder(task.getAction()));
+        if (tryAcquire()) {
+            statsHolder.current.inc();
+            final Runnable release = new RunOnce(() -> {
+                release();
+                statsHolder.current.dec();
+                statsHolder.completed.inc();
+            });
+            boolean success = false;
+            try {
+                executeAction.apply(task, request, ActionListener.runBefore(listener, release::run));
+                success = true;
+            } finally {
+                if (success == false) {
+                    release.run();
+                }
+            }
+        } else {
+            listener.onFailure(new EsRejectedExecutionException("too concurrent stats requests"));
```

Comment:
I think this is missing a word?
```suggestion
            listener.onFailure(new EsRejectedExecutionException("too many concurrent stats requests"));
```

Also, perhaps we can mention the limit in the error message, something like "too many concurrent stats requests (limit: 100)"?

Reply:
That could be helpful. I will add it.